streaming/base/batching/device_per_stream.py

```diff
@@ -76,10 +76,18 @@ def generate_work_device_per_stream_batching(dataset: StreamingDataset, world: W
             # same as the ratio of the stream's samples to overall samples.
             # This ensures that the overall training shuffle block size is still approximately
             # equal to what is set by the user, and allows for reasoning about cache_limit as well.
+            if samples_in_stream < dataset.num_canonical_nodes:  # third patch
+                logger.warning(
+                    f'Because of the `device_per_stream` batching method and stream with index {stream_id} '
+                    +
+                    f'has less samples ({samples_in_stream}) than canonical nodes ({dataset.num_canonical_nodes}), stream will be dropped.'
+                )
+                continue
+
             if not isinstance(dataset.shuffle_block_size, int):
                 raise TypeError(f'Dataset `shuffle_block_size` must be an integer. ' +
                                 f'Got {type(dataset.shuffle_block_size)} instead.')
-            shuffle_block_portion = int(dataset.shuffle_block_size * stream.proportion)
+            shuffle_block_portion = max(1, int(dataset.shuffle_block_size * stream.proportion))
```

Comment:
Not sure what the implication of this change would be, or why it is necessary?

Reply:
If you have a stream with few samples and a lot of samples across all streams, you end up with stream.proportion being very small and then `int(dataset.shuffle_block_size * stream.proportion) = 0`.
Then `get_shufffle` called next line will raise an error [this line](https://github.com/mosaicml/streaming/blob/ecca82006d6d1c97f7ee1d7c46bc5df61577efc9/streaming/base/shuffle/py1e.py#L101C61-L101C71) as it tries to sample an integer between in [0,0)

Reply:
I see, nice catch! In this case I would actually prefer to throw an error if a stream is not being used, so if `shuffle_block_portion == 0`, we should throw an error. A stream of data that's provided but not used in training is a significant issue and imo training should not proceed.